### PR TITLE
Ft/adding metrics

### PR DESCRIPTION
**Pull request by PNKDuy**
- Follow [here](https://github.com/kardiachain/go-kardia/blob/ft/adding_metrics/prometheus/README.md) for the configuration.
- Add `--metrics` arg to command line. The metric HTTP endpoint will be at HTTPhost:debugport/metrics.
- Change the prometheus.yml with the instruction at [here](https://github.com/kardiachain/go-kardia/blob/ft/adding_metrics/prometheus/README.md).
- Start prometheus service ( default port 9090).
- Add datasource in grafana.

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -377,6 +373,7 @@ func (c *Config) Start() {
 
 func (c *Config) StartDebug() error {
 	go func() {
+		log.Warn("Running router server")
```
Comment (trinhdn97) [inline cmd/main.go:376]: The log level should be `Info` or `Debug`

File: kai/state/cstate/execution.go [hunk 2]
```diff
@@ -124,8 +126,16 @@ func (blockExec *BlockExecutor) ApplyBlock(state LatestBlockState, blockID types
 		return state, block.Height(), fmt.Errorf("commit failed for application: %v", err)
 	}
 	state.AppHash = appHash
+	subStart := time.Now()
 	blockExec.store.Save(state)
 
+	if metrics.Enabled {
+		saveStateTimer.UpdateSince(subStart)
```
Comment (trinhdn97) [inline kai/state/cstate/execution.go:133]: Should we need timer evaluation metrics or storage metrics only @lewtran @thang14 ?

**Comment by lewtran** [comment]
Please add detailed descriptions in the PR; eg how to do the configurations, endpoints

File: mainchain/blockchain/block_operations.go [hunk 3]
```diff
@@ -140,17 +144,42 @@ func (bo *BlockOperations) CreateProposalBlock(
...
 	}
 
+	opStart := time.Now()
 	bo.saveBlockInfo(blockInfo, block)
 	bo.blockchain.DB().WriteHeadBlockHash(block.Hash())
 	bo.blockchain.DB().WriteTxLookupEntries(block)
 	bo.blockchain.DB().WriteAppHash(block.Height(), root)
 	bo.blockchain.InsertHeadBlock(block)
 
+	if metrics.Enabled {
```
Comment (trinhdn97) [inline mainchain/blockchain/block_operations.go:160]: These kinds of metrics would be considered as heavy operations. Should we separate these to be independent from other legacy metrics (rpc, trie, tx_pool,...) @lewtran @thang14 ?

File: prometheus/README.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+```
+./cmd --network devnet --genesis cfg/genesis_devnet.yaml --node cfg/devnet_node1.yaml --metrics
+```
+
+
+### Configuring Prometheus to monitor node
+Save the following Prometheus configuration as a file named prometheus.yml:
+```
+global:
+  scrape_interval:     15s # By default, scrape targets every 15 seconds.
```
Comment (trinhdn97) [inline prometheus/README.md:18]: The `scrape_interval` should be around 5s as our average blocktime.

File: prometheus/collector.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+import (
+	"bytes"
+	"fmt"
+	"github.com/kardiachain/go-kardia/lib/metrics"
+	"strconv"
+	"strings"
+)
+
+var (
+	helpTpl				   = "# HELP %s\n"
```
Comment (trinhdn97) [inline prometheus/collector.go:12]: Please run `go-fmt` all over the project to fix these linting nitpicks.

**Review by trinhdn97 — COMMENTED** [review]
Need some confirmations from @lewtran @thang14

File: kai/state/cstate/execution.go (hunk 2 above)
Comment (lewtran) [reply kai/state/cstate/execution.go:133]: this is ok for benchmarking purpose

**Pull request closed without merging**